### PR TITLE
Remove dependency of Clock on Prim and remove IORef.Prim

### PR DESCRIPTION
**Pull request by adithyaov**
```
Prelude.Serial(cpuTime)
Benchmark                                                                                                     default(μs)
------------------------------------------------------------------------------------------------------------- -----------
All.Prelude.Serial/o-1-space.generation.fromStorableIORef                                                           27.93
All.Prelude.Serial/o-1-space.generation.fromPrimIORef                                                               27.93

Prelude.Serial(Allocated)
Benchmark                                                                                                     default(Bytes)
------------------------------------------------------------------------------------------------------------- --------------
All.Prelude.Serial/o-1-space.generation.fromStorableIORef                                                               0.00
All.Prelude.Serial/o-1-space.generation.fromPrimIORef                                                                   0.00

Prelude.Serial(maxrss)
Benchmark                                                                                                     default(MiB)
------------------------------------------------------------------------------------------------------------- ------------
All.Prelude.Serial/o-1-space.generation.fromStorableIORef                                                             3.00
All.Prelude.Serial/o-1-space.generation.fromPrimIORef                                                                 3.00
```
```
Prelude.Serial(cpuTime)
Benchmark                                                                                                     default(0)(μs) default(1) - default(0)(%)
------------------------------------------------------------------------------------------------------------- -------------- --------------------------
All.Prelude.Serial/o-1-space.generation.absTimes                                                                       38.60                      +1.51
All.Prelude.Serial/o-1-space.mapping.pollCounts 1 second                                                              146.06                      +3.70
All.Prelude.Serial/o-1-space.mapping.tapRate 1 second                                                                 146.06                      +4.14

Prelude.Serial(Allocated)
Benchmark                                                                                                     default(0)(KiB) default(1) - default(0)(%)
------------------------------------------------------------------------------------------------------------- --------------- --------------------------
All.Prelude.Serial/o-1-space.generation.absTimes                                                                         1.49                      -0.46
All.Prelude.Serial/o-1-space.mapping.pollCounts 1 second                                                                 3.00                       0.00
All.Prelude.Serial/o-1-space.mapping.tapRate 1 second                                                                    3.00                       0.00
```

### CODE REVIEW

File: core/src/Streamly/Internal/Data/IORef/Storable.hs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import GHC.Exts ( MutableByteArray#, Ptr(..), RealWorld
+                , byteArrayContents#, newPinnedByteArray#, touch#
+                , unsafeCoerce#
+                )
+import GHC.Int (Int(..))
+import GHC.IO (IO(..))
+
+import qualified Streamly.Internal.Data.Stream.StreamD.Type as D
+
+-- XXX data IORef a = IORef (Ptr a) ?
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/IORef/Storable.hs:56]: NO!

File: core/src/Streamly/Internal/Data/IORef/Storable.hs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+--
+-- /Pre-release/
+{-# INLINE newIORef #-}
+newIORef :: forall a. Storable a => a -> IO (IORef a)
+newIORef x = do
+    ref <-
+        IO
+            (\s# ->
+                 let !(I# i#) = sizeOf (undefined :: a)
+                  in case newPinnedByteArray# i# s# of
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/IORef/Storable.hs:89]: Do not use Pinned memory.

File: core/src/Streamly/Internal/Data/IORef/Storable.hs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+-- | Modify the value of an 'IORef' using a function with strict application.
+--
+-- /Pre-release/
+{-# INLINE modifyIORef' #-}
+modifyIORef' :: Storable a => IORef a -> (a -> a) -> IO ()
+modifyIORef' ref g =
+    asPtrUnsafe ref $ \ptr -> do
+        a <- peekByteOff ptr 0
+        let a' = g a
+        a' `seq` pokeByteOff ptr 0 a'
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/IORef/Storable.hs:118]: `seq` is redundant here as we are writing the value to a bytearray it cannot be lazy.

File: core/src/Streamly/Internal/Data/IORef/Storable.hs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+-- | Read a value from an 'IORef'.
+--
+-- /Pre-release/
+{-# INLINE readIORef #-}
+readIORef :: Storable a => IORef a -> IO a
+readIORef ref = asPtrUnsafe ref $ \ptr -> peekByteOff ptr 0
+
+-- XXX What if GC + reallocation happens in the middle?
+-- XXX Will unpinned work?
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/IORef/Storable.hs:108]: We should use unpinned memory.

File: core/src/Streamly/Internal/Data/IORef/Storable.hs (hunk 3 above)
Comment (harendra-kumar) [reply core/src/Streamly/Internal/Data/IORef/Storable.hs:118]: In fact, we should rename this function to `modifyIORef` as it is strict by default and there is no lazy version.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: core/src/Streamly/Internal/Data/IORef/Storable.hs (hunk 4 above)
Comment (adithyaov) [reply core/src/Streamly/Internal/Data/IORef/Storable.hs:108]: Will unpinned memory work while playing with pointers?

File: core/src/Streamly/Internal/Data/IORef/Storable.hs (hunk 2 above)
Comment (adithyaov) [reply core/src/Streamly/Internal/Data/IORef/Storable.hs:89]: We need to use pinned memory. We extract the `Addr#` when working with `Storable`. Unpinned memory will be unsafe?

**Review by harendra-kumar — COMMENTED** [review]
Since we want to use unpinned memory, and we are going to use Prim for arrays as well for that reason, let's drop this change.

**Pull request closed without merging**